File: src/peated/models/notification.py
from sqlalchemy import (
    BigInteger,
    Boolean,
    Column,
    DateTime,
    Enum,
    ForeignKeyConstraint,
    Index,
    PrimaryKeyConstraint,
    text,
)

from peated.db.base import Base


class Notification(Base):
    __tablename__ = "notifications"
    __table_args__ = (
        ForeignKeyConstraint(["from_user_id"], ["user.id"], name="notifications_from_user_id_user_id_fk"),
        ForeignKeyConstraint(["user_id"], ["user.id"], name="notifications_user_id_user_id_fk"),
        PrimaryKeyConstraint("id", name="notifications_pkey"),
        Index(
            "notifications_unq",
            "user_id",
            "object_id",
            "object_type",
            "created_at",
            unique=True,
        ),
    )

    id = Column(BigInteger)
    user_id = Column(BigInteger, nullable=False)
    object_id = Column(BigInteger, nullable=False)
    object_type = Column(
        Enum(
            "bottle",
            "edition",
            "entity",
            "tasting",
            "toast",
            "follow",
            "comment",
            name="object_type",
        ),
        nullable=False,
    )
    created_at = Column(DateTime, nullable=False)
    read = Column(Boolean, nullable=False, server_default=text("false"))
    from_user_id = Column(BigInteger)

File: src/server/db/model/notification.go
package model

import (
	"time"

	"gorm.io/gorm"
)

type NotificationType string

const (
	NotificationTypeComment       NotificationType = "comment"
	NotificationTypeToast         NotificationType = "toast"
	NotificationTypeFriendRequest NotificationType = "friend_request"
)

type Notification struct {
	gorm.Model
	ID         uint64           `gorm:"primaryKey" json:"id"`
	UserID     uint64           `json:"user_id"`
	FromUserID uint64           `json:"from_user_id"`
	ObjectID   uint64           `json:"object_id"`
	Type       NotificationType `json:"type"`

	CreatedByID uint64    `json:"created_by_id"`
	CreatedAt   time.Time `json:"created_at"`

	User     User
	FromUser User
}

func (Notification) TableName() string {
	return "notifications"
}

type Notifications []*Notification
